include/mbedtls/pk.h

```diff
@@ -104,7 +104,37 @@ typedef struct mbedtls_pk_rsassa_pss_options
 /**
  * \brief           Maximum size of a signature made by mbedtls_pk_sign().
  */
+/* This fallback value is used if there is no software signature support.
+ * This is possible even if check_config.h is included, for example if
+ * MBEDTLS_ECDH_C is enabled but neither MBEDTLS_ECDSA_C  nor MBEDTLS_RSA_C.
+ * Use MBEDTLS_MPI_MAX_SIZE which is the maximum size than an RSA-alt
+ * implementation can produce, assuming that MBEDTLS_MPI_MAX_SIZE is set
+ * correctly. This is not necessarily the best choice of size and it may
+ * change in future versions. */
 #define MBEDTLS_PK_SIGNATURE_MAX_SIZE MBEDTLS_MPI_MAX_SIZE
+#if defined(MBEDTLS_RSA_C) &&                                   \
+    MBEDTLS_MPI_MAX_SIZE > MBEDTLS_PK_SIGNATURE_MAX_SIZE
+#undef MBEDTLS_PK_SIGNATURE_MAX_SIZE
+#define MBEDTLS_PK_SIGNATURE_MAX_SIZE MBEDTLS_MPI_MAX_SIZE
```

Comment:
These lines appear to unset `MBEDTLS_PK_SIGNATURE_MAX_SIZE` and then reset it back to `MBEDTLS_MPI_MAX_SIZE`. Am I missing something here?

Reply:
It's a coincidence. The undef/define gymnastic is necessary to redefine the value. While you can express max(…) using C operators, it doesn't scale. It's still ok for max(x1,x2) (`x1 > x2 ? x1 : x2`) but it quickly grows unmanageable for `max(x1, x2, x3, …)`.

I'll add a comment to explain what's going on.

Reply:
Ah, but no, the code isn't working as intended! It's always setting `MBEDTLS_PK_SIGNATURE_MAX_SIZE` to be at least `MBEDTLS_MPI_MAX_SIZE`, which I wanted to avoid in the fairly common case where `MBEDTLS_MPI_MAX_SIZE` is kept to a large value but RSA signature is disabled and `MBEDTLS_PK_SIGNATURE_MAX_SIZE` encompasses ECDSA only with a much smaller maximum size.

Reply:
Uh, actually, this fallback _was_ working as intended, sort of. It's not a fallback in the sense of “if nothing else is defined”, but something that can arise even if nothing is defined.

Except that the RSA_ALT case can't happen if RSA_ALT is not enabled, so that part was wrong.

Reply:
Fixed.